layouts/shortcodes/resources_common_compile_begin.md

```diff
@@ -2,65 +2,124 @@ Use `.run_action(:some_action)` at the end of a resource block to run
 ```
 
 where `action` is set to `:nothing` to ensure the `run_action` is run
 during the compile phase and not later during the execution phase.
 
+This can be simplified by using the `compile_time` flag in Chef Infra
+Client 16 and later versions:
+
+``` ruby
+build_essential 'Install compilers' do
+  compile_time true
+end
+```
+
+That flag both forces the resource to run at compile time and sets the
+converge action to :nothing.
+
 The following examples show when (and when not) to use `run_action`.
 
-**Update a package cache**
+**Using Custom Resources preferred to forcing to compile time**
+
+Compile time execution is often used to install gems before requiring
+them in recipe code.
 
-Sometimes it is necessary to ensure that an operating system's package
-cache is up to date before installing packages. For example, on Debian
-or Ubuntu systems, the Apt cache should be updated:
+This is a poor pattern, since gems may depend on native gems, which
+may result in needing to install compilers at compile time.
 
 ``` ruby
-if node['apt']['compile_time_update'] && ( !::File.exist?('/var/lib/apt/periodic/update-success-stamp') || !::File.exist?(first_run_file) )
-  e = bash 'apt-get-update at compile time' do
-    code <<-EOH
-      apt-get update
-      touch #{first_run_file}
-    EOH
-    ignore_failure true
-    only_if { apt_installed? }
-    action :nothing
-  end
-  e.run_action(:run)
+build_essential 'Install compilers' do
+  compile_time true
 end
-```
 
-where `e.run_action(:run)` tells Chef Infra Client to run the
-`apt-get update` command during the compile phase. This example can be
-found in the `default.rb` recipe of the [apt
-cookbook](https://github.com/chef-cookbooks/apt) that is maintained by
-Chef.
+chef_gem 'aws-dsk' do
+  compile_time true
+end
 
-**Use the chef_gem resource for Ruby gems**
+require 'aws-sdk'
+```
 
-A very common use case is to install a gem during the compile phase so
-that it will be available to Chef Infra Client during the execution
-phase. This is why the **chef_gem** resource exists. For example, this:
+A better strategy is to move the code which requires the gem into
+a custom resource.  Since all the actions of custom resources run
+at converge time, this has the effect of deferring the require of
+the gem to later in the overall chef-client execution.  Unified
+mode can also be used in the resource to eliminate compile/converge
+mode issues entirely:
 
 ``` ruby
-chef_gem 'foo' do
-  action :install
+unified_mode true
+
+action :run do
+  build_essential 'Install compilers'
+
+  chef_gem 'aws-sdk'
+
+  require 'aws-sdk'
 end
 ```
 
-is effectively the same as
+**Download and parse a configuration file**
+
+A common use case is to download a configuration file, parse it and then
+use the values in templates and to control other configuration.
+
+And important distinction to make is that the downloaded configuration file
+is being used internally by the Chef Infra Client as temporary state.
```

Comment:
Not sure of the context here but maybe one of these suggestions: 

```suggestion
An important distinction to make is that the downloaded configuration file
is being used internally by the Chef Infra Client in a temporary state.
```
Or maybe:

`used internally while the Chef Infra Client is in a temporary state.`

Reply:
definitely not the second one.

Reply:
So the file that is being downloaded does not really affect the running system or the applications.  It isn't a file in /etc it isn't some kind of O/S config, it isn't an application config file.  So the file is downloaded, then recipe reads it, then the chef-client goes and does the REAL work of configuring the system.  So it is internal or temporary or transient state that is only consumed by chef-client itself.

Dunno if that helps.